apstra/api_custom_collector.go

```diff
@@ -0,0 +1,257 @@
+package apstra
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+)
+
+const (
+	apiUrlCollectors              = "/api/telemetry/collectors"
+	apiUrlCollectorsByServiceName = apiUrlCollectors + apiUrlPathDelim + "%s"
+)
+
+//var (
+//	_ json.Marshaler   = (*Collector)(nil)
+//	_ json.Unmarshaler = (*Collector)(nil)
+//)
+
+type CollectorPlatform struct {
+	OsType    CollectorOSType
+	OsVersion CollectorOSVersion
+	OsFamily  []CollectorOSVariant
+	Model     string
+}
+
+func (o *CollectorPlatform) UnmarshalJSON(data []byte) error {
+	var raw struct {
+		OsType    string `json:"os_type"`
+		OsVersion string `json:"os_version"`
+		OsFamily  string `json:"family"`
+		Model     string `json:"model"`
+	}
+
+	err := json.Unmarshal(data, &raw)
+	if err != nil {
+		return err
+	}
+
+	err = o.OsType.FromString(raw.OsType)
+	if err != nil {
+		return err
+	}
+
+	err = o.OsVersion.FromString(raw.OsVersion)
+	if err != nil {
+		return err
+	}
+
+	o.Model = raw.Model
+
+	for _, v := range strings.Split(raw.OsFamily, ",") {
+		var variant CollectorOSVariant
+		err = variant.FromString(v)
+		if err != nil {
+			return err
+		}
+		o.OsFamily = append(o.OsFamily, variant)
+	}
+
+	return nil
+}
+
+func (o *CollectorPlatform) MarshalJSON() ([]byte, error) {
+	var raw struct {
+		OsType    string `json:"os_type"`
+		OsVersion string `json:"os_version"`
+		OsFamily  string `json:"family"`
+		Model     string `json:"model"`
+	}
+	raw.OsType = o.OsType.String()
+	raw.OsVersion = o.OsVersion.String()
+	raw.Model = o.Model
+	raw.OsFamily = o.OsFamily[0].String()
+	for _, v := range o.OsFamily[1:] {
+		raw.OsFamily = raw.OsFamily + "," + v.String()
+	}
+	return json.Marshal(raw)
+}
+
+type Query struct {
```

Comment:
"Query" is a little vague for a public object. Could this be a "ServiceQuery" or a "CollectorQuery" or something?

Reply:
Renamed to CollectorQuery

Reply:
Renamed to CollectorQuery

Reply:
renamed to collectorquery